cpp/src/parquet/page_index_test.cc

```diff
@@ -416,4 +416,420 @@ TEST(PageIndex, DeterminePageIndexRangesInRowGroupWithMissingPageIndex) {
+                offset_index->page_locations()[0].offset);
+      ASSERT_EQ(page_locations[i].compressed_page_size,
+                offset_index->page_locations()[0].compressed_page_size);
+      ASSERT_EQ(page_locations[i].first_row_index,
+                offset_index->page_locations()[0].first_row_index);
+    } else {
+      ASSERT_FALSE(column_index_locations[i].has_value());
+      ASSERT_FALSE(offset_index_locations[i].has_value());
+    }
+  }
+}
+
+TEST(PageIndex, TestPageIndexBuilderWithTwoRowGroups) {
+  schema::NodePtr root = schema::GroupNode::Make(
+      "schema", Repetition::REPEATED, {schema::ByteArray("c1"), schema::ByteArray("c2")});
+  SchemaDescriptor schema;
+  schema.Init(root);
+
+  // Prepare page stats and page locations for two row groups.
+  const std::vector<std::vector<EncodedStatistics>> page_stats = {
+      /* 1st row group */
+      {EncodedStatistics().set_min("a").set_max("b"),
+       EncodedStatistics().set_null_count(0).set_min("A").set_max("B")},
+      /* 2nd row group */
+      {EncodedStatistics() /* corrupted stats */,
+       EncodedStatistics().set_null_count(0).set_min("bar").set_max("foo")}};
+  const std::vector<std::vector<PageLocation>> page_locations = {
+      /* 1st row group */
+      {{/*offset=*/128, /*compressed_page_size=*/512,
+        /*first_row_index=*/0},
+       {/*offset=*/1024, /*compressed_page_size=*/512,
+        /*first_row_index=*/0}},
+      /* 2nd row group */
+      {{/*offset=*/128, /*compressed_page_size=*/512,
+        /*first_row_index=*/0},
+       {/*offset=*/1024, /*compressed_page_size=*/512,
+        /*first_row_index=*/0}}};
+  const std::vector<int64_t> final_positions = {1024, 2048};
+
+  // Create builder and add pages of two row groups.
+  const size_t num_row_groups = 2;
+  const size_t num_columns = 2;
+  const size_t num_pages = 1;
+  auto builder = PageIndexBuilder::Make(&schema);
+  for (size_t rg = 0; rg < num_row_groups; ++rg) {
+    ASSERT_NO_THROW(builder->AppendRowGroup());
+    for (int c = 0; c < static_cast<int>(num_columns); ++c) {
+      ASSERT_NO_THROW(builder->GetColumnIndexBuilder(c)->AddPage(page_stats[rg][c]));
+      ASSERT_NO_THROW(builder->GetColumnIndexBuilder(c)->Finish());
+      ASSERT_NO_THROW(builder->GetOffsetIndexBuilder(c)->AddPage(page_locations[rg][c]));
+      ASSERT_NO_THROW(builder->GetOffsetIndexBuilder(c)->Finish(final_positions[rg]));
+    }
+  }
+  ASSERT_NO_THROW(builder->Finish());
+
+  // Verify WriteTo only serializes valid page index.
+  auto sink = CreateOutputStream();
+  PageIndexLocation location;
+  builder->WriteTo(sink.get(), &location);
+  PARQUET_ASSIGN_OR_THROW(auto buffer, sink->Finish());
+  ASSERT_EQ(num_row_groups, location.column_index_location.size());
+  ASSERT_EQ(num_row_groups, location.offset_index_location.size());
+
+  // Verify data of deserialized page index.
+  auto properties = default_reader_properties();
+  for (size_t rg = 0; rg < num_row_groups; ++rg) {
```

Comment:
would unwinding the loops + appropriate helper methods make this signifantly more verbose?

Having logic in tests, at least for me, makes them much more difficult to read (sometimes it is warranted but sometimes it is cleaner to try make things less complex).

Reply:
I have rewritten the tests. Please take a look, thanks! @emkornfield